File: BenKim/2022-09-13-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        # 아무리 큰 금액도 갱신되게 하기 위해 os의 가장큰값을 최소가격으로 설정
        min_price = sys.maxsize
        profit = 0
        
        # 최대가격과 최소가격을 갱신
        for price in prices:
            min_price = min(price, min_price)
            profit = max(profit, price - min_price)
        return profit

File: DongwanKim/2022-09-13-BestTimeToBuyAndSellStock.js
/**
 * You are given an array prices where prices[i] is the price of a given stock on the ith day.

You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.



Example 1:

Input: prices = [7,1,5,3,6,4]
Output: 5
Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
Note that buying on day 2 and selling on day 1 is not allowed because you must buy before you sell.
Example 2:

Input: prices = [7,6,4,3,1]
Output: 0
Explanation: In this case, no transactions are done and the max profit = 0.


Constraints:

1 <= prices.length <= 105
0 <= prices[i] <= 104
 */

// first solution - 2 pointers
var maxProfit = function (prices) {
  let profit = 0;
  let left = 0;
  let right = 1;

  //buy low sell high
  while (right < prices.length) {
    // if profitable
    if (prices[left] < prices[right]) {
      let currProfit = prices[right] - prices[left];
      profit = Math.max(currProfit, profit);
    } else {
      // if left price is bigger than right price, move left pointer to right
      // since we only want to buy at lowest price
      left = right;
    }
    // move right pointer to next day to check if its profitable
    right++;
  }
  return profit;
  //test
};
